### PR TITLE
Measure latencies for two fixture wallets

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#1097 

# Overview

- [x] I have measured latencies for all specified in U/S endpoints 
- [x] I have used iohk-monitoring-framework primitives to do measurement
- [x] I have added structural logging ServerLog
- [x] I have defined measureLatencies inside requests 
- [x] I have updated LoggingSpec in core

# Comments

### CODE REVIEW

**Comment by rvl** [comment]
This is a good start.

The important thing about Criterion is not so much how it measures time but that it repeats the benchmark to get an accurate result, and calculates the statistical distribution of results. If we only measure once, we don't get that information.

If this benchmark is run as part of a test suite, then the results are going to be inaccurrate, because other test cases may be running at the same time. So it should be a separate benchmark program. Then it can be run independently without the results being skewed by other processes.

In any case, I think we should be using the iohk-monitoring framework for these type of benchmarks. It was written specifically for this purpose. Log traces can have timestamps attached - so just run the tests and collect the timings from the logs.

**Comment by paweljakubas** [comment]
bors try

**Comment by paweljakubas** [comment]
@rvl I have followed the path we discussed privately. So defined high level `ServerLog`, relocated everything to `jormungandr/test/bench`, used here `traceinTVar` and precise pattern matching (and clearing the tvar in between). And then relied on meta fragment of LogObject to get UTCTime. Also did some commit clearing (removed first 4 commits from history - those that added integration tests). And it works. Just we need to do something like this : 
```
$ stack bench cardano-wallet-jormungandr
cardano-wallet-jormungandr> benchmarks  
Running 1 benchmarks...                 
Benchmark latency: RUNNING...           
Latencies for two fixture wallets scenario
    listWallets        - 20.2 ms        
    getWallet          - 11.1 ms        
    getUTxOsStatistics - 1.9 ms         
    listAddresses      - 7.1 ms         
    listTransactions   - 2.5 ms         
    postTransactionFee - 1.7 ms         
    listStakePools     - 1.2 ms         
    getNetworkInfo     - 0.4 ms         
Benchmark latency: FINISH
```

**Comment by paweljakubas** [comment]
bors try

**Comment by paweljakubas** [comment]
bors try

**Comment by paweljakubas** [comment]
bors try-

**Comment by paweljakubas** [comment]
bors try

File: lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs [hunk 1]
```diff
@@ -111,80 +119,114 @@ spec = describe "Logging Middleware"
...
+            , (Debug, "")
+            , (Debug, "LogRequestFinish")
             ]
 
     it "different request ids" $ \ctx -> property $ \(NumberOfRequests n) ->
         monadicIO $ liftIO $ do
             void $ mapConcurrently (const (get ctx "/get")) (replicate n ())
             entries <- readTVarIO (logs ctx)
-            let index = Map.fromList [ (loName l, l) | l <- entries ]
+            let getReqId l = case loContent l of
```
Comment (rvl) [inline lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs:222]: This could still be a list comprehension, right? Failed pattern matches are filtered out.

File: lib/jormungandr/test/integration/Cardano/Faucet.hs [hunk 2]
```diff
@@ -1112,3 +1126,31 @@ genFaucets file n = do
             [ paymentAddress @'Testnet (publicKey $ addrXPrv ix)
             | ix <- [minBound..maxBound]
             ]
+
+sockAddrPort :: SockAddr -> Port a
```
Comment (rvl) [inline lib/jormungandr/test/integration/Cardano/Faucet.hs:1121]: Why is this moved to the Faucet module?

**Review by rvl — COMMENTED** [review]
I moved `ServerLog` into `Cardano.Wallet.Jormungandr`, which is a more natural place for it.

The changes to LoggingSpec removed assertions about log message formatting and hiding sensitive data. I added these back because they are important to test.

The benchmark was only doing a single measurement. This won't be very accurate for the times that are being measured here. Slightly better is running multiple times and taking an average. I modified the benchmark to do that.

I have a couple questions/comments...

**Comment by rvl** [comment]
Also when we say "latency", we need to say latency of _what_. We are measuring wallet server API response time.

**Review by rvl — APPROVED** [review]

File: lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs:222]: my intention here was to get RequestId from each log, and then nubBy the entries by this requestId. Thanks to that, I am expected to end up with unique logs having distinct RequestId. So...probably can be list comprehension, but I find the proposed version more clear (but this is subjective;-)

File: lib/jormungandr/test/integration/Cardano/Faucet.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/jormungandr/test/integration/Cardano/Faucet.hs:1121]: I moved `sockAddrPort` and `mkFeeEstimator` to Faucet as they are reused in `test/bench/Latency.hs` and `test/integration/Main.hs` and wanted to avoid duplication

File: lib/jormungandr/test/bench/Latency.hs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+                faucet <- initFaucet
+                putMVar ctx $ Context
+                    { _cleanup = pure ()
+                    , _manager = manager
+                    , _nodePort = nPort
+                    , _walletPort = sockAddrPort wAddr
+                    , _faucet = faucet
+                    , _feeEstimator = mkFeeEstimator (getFeePolicy bp)
+                    , _target = Proxy
+                    }
```
Comment (KtorZ) [inline lib/jormungandr/test/bench/Latency.hs:292]: Commenting after seeing another comment below about "re-using" the fee estimator:

I don't think there's any use for the fee estimator in this latency benchmark. This function is used from within a few integration scenario to test edge-cases about the fee policy. The context for the benchmark is much simpler and there's no need for this fee estimator at all.

cc @paweljakubas @rvl

Comment (paweljakubas) [reply lib/jormungandr/test/bench/Latency.hs:292]: addressed in https://github.com/input-output-hk/cardano-wallet/pull/1105/commits/4c12755758ffcb713bb9ea4c9a3200de167afaee
